### ISSUE TITLE
[🐛] Coding Challenge: Forged Review

### ISSUE THREAD

**Issue by kuljotbiring** [juice-shop/juice-shop#2389]
# :bug: Bug report

## Description

Coding Challenge: Forged Review
When working on this challenge, the following code is presented:

```
module.exports = function productReviews () {
  return (req: Request, res: Response, next: NextFunction) => {
    const user = security.authenticatedUsers.from(req)
    db.reviewsCollection.update(
      { _id: req.body.id },
      { $set: { message: req.body.message } },
      { multi: true }
    ).then(
      (result: { modified: number, original: Array<{ author: any }> }) => {
        res.json(result)
      }, (err: unknown) => {
        res.status(500).json(err)
      })
  }
}
```

The solutions state that the code should be ammended as follows:

```
module.exports = function productReviews () {
  return (req: Request, res: Response, next: NextFunction) => {
    const user = security.authenticatedUsers.from(req)
    db.reviewsCollection.update(
      { _id: req.body.id, author: user.data.email },
       {$set: { message: req.body.message, author: user.data.email } },
      // { multi: true }
    ).then(
      (result: { modified: number, original: Array<{ author: any }> }) => {
        res.json(result)
      }, (err: unknown) => {
        res.status(500).json(err)
      })
  }
}
```
We can see lines 5-7 have been updated. However, incorporating these changes does not patch the vulnerability as hosting the patched version on Heroku still allows me to use Burp and modify the review's author as admin and leave a forged comment.

### Is this a regression?

N/A

## :microscope: Minimal Reproduction

1. Make the recommended code changes above and host the application.
2. Open Burp Suite Proxy and use the browser to navigate to the Juice Shop page 
3. Create a new user and login as that user
4. Now turn on intercept on Intercept in Burp Proxy
5. Create a comment and forward until you see the PUT request making the comment
6. Modify the author in the PUT request to be admin@juice-sh.op
7. Now forward the request
8. The comment will still be forged as the admin

## :fire: Exception or Error

<pre><code>

</code>N/A</pre>

## :deciduous_tree: Your Environment

<pre><code>

</code>Run on Heroku - not locally</pre>

### Additional Information

This is done using Burp Suite Proxy and its browser (Chromium). The application is hosted on Heroku

**Comment by bkimminich** [comment]
Thank you for reporting this in such detail! 😄👍 We'll look into it and see if we can come up with another fix recommendation.

**Comment by malikrohail** [comment]
can u assign this to me

**Comment by kuljotbiring** [comment]
has this been assigned to anyone yet or any progress?

**Comment by J12934** [comment]
nope. we rarely assign tickets to users as this it rarely works and blocks other people from starting. If the people then run out of time / or loose focus we have to chase them down which is really annoying for everybody.

So if somebody wants to take this up, feel free to just do so. 
If you started but haven't completed it yet feel free to open up a draft PR.

**Comment by ZaidArif47** [comment]
The vulnerability lies in the createProductReviews.ts file, where the author is set to `req.body.author` instead of `user.data.email`. Similarly, in the updateProductReviews.ts file, the update query only checks `{ _id: req.body.id }`, leaving the author field open to manipulation.

The fix is implemented in forgedReviewChallenge_2_correct.ts, where the author is added to the update query: `{ _id: req.body.id, author: user.data.email }`, ensuring that the review can only be updated by the authenticated user.

@kuljotbiring The solution that you have mentioned does not look like the right one to me. forgedReviewChallenge_2_correct.ts is the right solution. Pls check again and correct me if I'm wrong

**Comment by bkimminich** [comment]
> The vulnerability lies in the createProductReviews.ts file, where the author is set to `req.body.author` instead of `user.data.email`. Similarly, in the updateProductReviews.ts file, the update query only checks `{ _id: req.body.id }`, leaving the author field open to manipulation.
> 
> The fix is implemented in forgedReviewChallenge_2_correct.ts, where the author is added to the update query: `{ _id: req.body.id, author: user.data.email }`, ensuring that the review can only be updated by the authenticated user.
> 
> @kuljotbiring The solution that you have mentioned does not look like the right one to me. forgedReviewChallenge_2_correct.ts is the right solution. Pls check again and correct me if I'm wrong

Hm, yeah, the author from the PUT request wouldn't even be used anywhere in the actually fixed code. @kuljotbiring, did you by any chance just change the Typescript but then not build the application again, so it effectively still used the vulnerable transpiled JavaScript from before?
